### PR TITLE
Configure `ruff format`, use it instead of `black`

### PR DESCRIPTION
**Pull request by deepyaman**
> **NOTE:** Kedro datasets are moving from `kedro.extras.datasets` to a separate `kedro-datasets` package in
> [`kedro-plugins` repository](https://github.com/kedro-org/kedro-plugins). Any changes to the dataset implementations
> should be done by opening a pull request in that repository.
## Development notes

It's worth noting that there are a lot of warnings emitted by ruff (introduce an error on purpose, and run `make lint` to see them):

```
warning: Invalid `# noqa` directive on kedro/framework/session/session.py:53: expected a comma-separated list of codes (e.g., `# noqa: F401, F841`).
warning: Invalid `# noqa` directive on kedro/framework/session/session.py:81: expected a comma-separated list of codes (e.g., `# noqa: F401, F841`).
warning: Invalid `# noqa` directive on kedro/framework/session/session.py:189: expected a comma-separated list of codes (e.g., `# noqa: F401, F841`).
warning: Invalid `# noqa` directive on kedro/framework/session/session.py:420: expected a comma-separated list of codes (e.g., `# noqa: F401, F841`).
warning: Invalid `# noqa` directive on kedro/framework/cli/micropkg.py:108: expected a comma-separated list of codes (e.g., `# noqa: F401, F841`).
warning: Invalid `# noqa` directive on kedro/framework/cli/micropkg.py:117: expected a comma-separated list of codes (e.g., `# noqa: F401, F841`).
warning: Invalid `# noqa` directive on kedro/framework/cli/micropkg.py:261: expected a comma-separated list of codes (e.g., `# noqa: F401, F841`).
warning: Invalid `# noqa` directive on kedro/framework/cli/micropkg.py:285: expected a comma-separated list of codes (e.g., `# noqa: F401, F841`).
warning: Invalid `# noqa` directive on kedro/framework/cli/micropkg.py:368: expected a comma-separated list of codes (e.g., `# noqa: F401, F841`).
warning: Invalid `# noqa` directive on kedro/framework/cli/micropkg.py:379: expected a comma-separated list of codes (e.g., `# noqa: F401, F841`).
warning: Invalid `# noqa` directive on kedro/framework/cli/micropkg.py:397: expected a comma-separated list of codes (e.g., `# noqa: F401, F841`).
warning: Invalid `# noqa` directive on kedro/framework/cli/hooks/manager.py:2: expected a comma-separated list of codes (e.g., `# noqa: F401, F841`).
warning: Invalid `# noqa` directive on kedro/framework/cli/pipeline.py:68: expected a comma-separated list of codes (e.g., `# noqa: F401, F841`).
warning: Invalid `# noqa` directive on kedro/framework/cli/pipeline.py:74: expected a comma-separated list of codes (e.g., `# noqa: F401, F841`).
warning: Invalid `# noqa` directive on kedro/framework/cli/pipeline.py:103: expected a comma-separated list of codes (e.g., `# noqa: F401, F841`).
warning: Invalid `# noqa` directive on kedro/framework/cli/pipeline.py:143: expected a comma-separated list of codes (e.g., `# noqa: F401, F841`).
warning: Invalid `# noqa` directive on kedro/framework/cli/pipeline.py:211: expected a comma-separated list of codes (e.g., `# noqa: F401, F841`).
warning: Invalid `# noqa` directive on kedro/io/data_catalog.py:432: expected a comma-separated list of codes (e.g., `# noqa: F401, F841`).
warning: Invalid `# noqa` directive on kedro/config/omegaconf_config.py:284: expected a comma-separated list of codes (e.g., `# noqa: F401, F841`).
warning: Invalid `# noqa` directive on kedro/runner/parallel_runner.py:96: expected a comma-separated list of codes (e.g., `# noqa: F401, F841`).
warning: Invalid `# noqa` directive on kedro/runner/parallel_runner.py:100: expected a comma-separated list of codes (e.g., `# noqa: F401, F841`).
warning: Invalid `# noqa` directive on kedro/runner/parallel_runner.py:168: expected a comma-separated list of codes (e.g., `# noqa: F401, F841`).
warning: Invalid `# noqa` directive on kedro/runner/parallel_runner.py:226: expected a comma-separated list of codes (e.g., `# noqa: F401, F841`).
warning: Invalid `# noqa` directive on kedro/runner/parallel_runner.py:279: expected a comma-separated list of codes (e.g., `# noqa: F401, F841`).
warning: Invalid `# noqa` directive on kedro/runner/parallel_runner.py:302: expected a comma-separated list of codes (e.g., `# noqa: F401, F841`).
warning: Invalid `# noqa` directive on kedro/framework/cli/project.py:71: expected a comma-separated list of codes (e.g., `# noqa: F401, F841`).
warning: Invalid `# noqa` directive on kedro/framework/cli/project.py:79: expected a comma-separated list of codes (e.g., `# noqa: F401, F841`).
warning: Invalid `# noqa` directive on kedro/framework/cli/project.py:102: expected a comma-separated list of codes (e.g., `# noqa: F401, F841`).
warning: Invalid `# noqa` directive on kedro/framework/cli/project.py:135: expected a comma-separated list of codes (e.g., `# noqa: F401, F841`).
warning: Invalid `# noqa` directive on kedro/framework/cli/project.py:235: expected a comma-separated list of codes (e.g., `# noqa: F401, F841`).
warning: Invalid `# noqa` directive on kedro/framework/cli/project.py:278: expected a comma-separated list of codes (e.g., `# noqa: F401, F841`).
warning: Invalid `# noqa` directive on kedro/framework/context/context.py:134: expected a comma-separated list of codes (e.g., `# noqa: F401, F841`).
warning: Invalid `# noqa` directive on kedro/pipeline/modular_pipeline.py:215: expected a comma-separated list of codes (e.g., `# noqa: F401, F841`).
warning: Invalid `# noqa` directive on kedro/io/partitioned_dataset.py:41: expected a comma-separated list of codes (e.g., `# noqa: F401, F841`).
warning: Invalid `# noqa` directive on kedro/io/partitioned_dataset.py:195: expected a comma-separated list of codes (e.g., `# noqa: F401, F841`).
warning: Invalid `# noqa` directive on kedro/io/partitioned_dataset.py:257: expected a comma-separated list of codes (e.g., `# noqa: F401, F841`).
warning: Invalid `# noqa` directive on kedro/io/partitioned_dataset.py:513: expected a comma-separated list of codes (e.g., `# noqa: F401, F841`).
warning: Invalid `# noqa` directive on kedro/runner/runner.py:289: expected a comma-separated list of codes (e.g., `# noqa: F401, F841`).
warning: Invalid `# noqa` directive on kedro/ipython/__init__.py:142: expected a comma-separated list of codes (e.g., `# noqa: F401, F841`).
warning: Invalid `# noqa` directive on kedro/ipython/__init__.py:153: expected a comma-separated list of codes (e.g., `# noqa: F401, F841`).
warning: Invalid `# noqa` directive on kedro/config/common.py:127: expected a comma-separated list of codes (e.g., `# noqa: F401, F841`).
warning: Invalid `# noqa` directive on kedro/framework/cli/utils.py:229: expected a comma-separated list of codes (e.g., `# noqa: F401, F841`).
warning: Invalid `# noqa` directive on kedro/framework/cli/utils.py:265: expected a comma-separated list of codes (e.g., `# noqa: F401, F841`).
warning: Invalid `# noqa` directive on kedro/framework/cli/utils.py:291: expected a comma-separated list of codes (e.g., `# noqa: F401, F841`).
warning: Invalid `# noqa` directive on kedro/framework/cli/utils.py:296: expected a comma-separated list of codes (e.g., `# noqa: F401, F841`).
warning: Invalid `# noqa` directive on kedro/framework/cli/utils.py:355: expected a comma-separated list of codes (e.g., `# noqa: F401, F841`).
warning: Invalid `# noqa` directive on kedro/framework/cli/utils.py:361: expected a comma-separated list of codes (e.g., `# noqa: F401, F841`).
warning: Invalid `# noqa` directive on kedro/framework/cli/utils.py:393: expected a comma-separated list of codes (e.g., `# noqa: F401, F841`).
warning: Invalid `# noqa` directive on kedro/framework/cli/utils.py:399: expected a comma-separated list of codes (e.g., `# noqa: F401, F841`).
warning: Invalid `# noqa` directive on kedro/framework/cli/cli.py:164: expected a comma-separated list of codes (e.g., `# noqa: F401, F841`).
warning: Invalid `# noqa` directive on kedro/pipeline/pipeline.py:74: expected a comma-separated list of codes (e.g., `# noqa: F401, F841`).
warning: Invalid `# noqa` directive on kedro/framework/project/__init__.py:3: expected a comma-separated list of codes (e.g., `# noqa: F401, F841`).
warning: Invalid `# noqa` directive on kedro/framework/project/__init__.py:134: expected a comma-separated list of codes (e.g., `# noqa: F401, F841`).
warning: Invalid `# noqa` directive on kedro/framework/project/__init__.py:211: expected a comma-separated list of codes (e.g., `# noqa: F401, F841`).
warning: Invalid `# noqa` directive on kedro/framework/project/__init__.py:332: expected a comma-separated list of codes (e.g., `# noqa: F401, F841`).
warning: Invalid `# noqa` directive on kedro/extras/datasets/tensorflow/tensorflow_model_dataset.py:139: expected a comma-separated list of codes (e.g., `# noqa: F401, F841`).
warning: Invalid `# noqa` directive on kedro/extras/datasets/tensorflow/tensorflow_model_dataset.py:160: expected a comma-separated list of codes (e.g., `# noqa: F401, F841`).
```

All of these `noqa` directives aren't getting parsed, so they're essentially useless. I haven't tried to clean this up in this PR.

## Developer Certificate of Origin
We need all contributions to comply with the [Developer Certificate of Origin (DCO)](https://developercertificate.org/). All commits must be signed off by including a `Signed-off-by` line in the commit message. [See our wiki for guidance](https://github.com/kedro-org/kedro/wiki/Guidelines-for-contributing-developers/).

If your PR is blocked due to unsigned commits, then you must follow the instructions under "Rebase the branch" on the GitHub Checks page for your PR. This will retroactively add the sign-off to all unsigned commits and allow the DCO check to pass.

## Checklist

- [x] Read the [contributing](https://github.com/kedro-org/kedro/blob/main/CONTRIBUTING.md) guidelines
- [x] Signed off each commit with a [Developer Certificate of Origin (DCO)](https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/managing-repository-settings/managing-the-commit-signoff-policy-for-your-repository)
- [ ] Opened this PR as a 'Draft Pull Request' if it is work-in-progress
- [ ] Updated the documentation to reflect the code changes
- [ ] Added a description of this change in the [`RELEASE.md`](https://github.com/kedro-org/kedro/blob/main/RELEASE.md) file
- [ ] Added tests to cover my changes
- [ ] Checked if this change will affect Kedro-Viz, and if so, communicated that with the Viz team

### CODE REVIEW

**Comment by deepyaman** [comment]
> Some seemingly legitimate failures in the `lint` job?

Taking a look; I didn't realize the 3.7 job was actually running before.

**Comment by deepyaman** [comment]
> Some seemingly legitimate failures in the `lint` job?

Fixed! Guess the old version of Ruff wasn't catching most of these?

**Comment by deepyaman** [comment]
> > Some seemingly legitimate failures in the `lint` job?
> 
> Taking a look; I didn't realize the 3.7 job was actually running before.



> Some seemingly legitimate failures in the `lint` job?

@astrojuanlu Fixed! Guess the old version of Ruff wasn't catching most of these?

File: tests/framework/context/test_context.py [hunk 1]
```diff
@@ -187,9 +187,7 @@ def extra_params(request):
 
 
 @pytest.fixture
-def dummy_context(
-    tmp_path, prepare_project_dir, env, extra_params
-):  # pylint: disable=unused-argument
+def dummy_context(tmp_path, prepare_project_dir, env, extra_params):  # pylint: disable=unused-argument
```
Comment (deepyaman) [inline tests/framework/context/test_context.py:190]:
```suggestion
def dummy_context(tmp_path, prepare_project_dir, env, extra_params):
```

**Comment by Galileo-Galilei** [comment]
(Seem that I am not allowed to review, but it looks ok for me). I guess we want to keep the documentation change out of this PR? The are still a lot of mentions of [black all across the codebase](https://github.com/search?q=repo%3Akedro-org%2Fkedro%20black&type=code). Some questions : 
- we want to change the documentation, but I guess this is out of scope for this PR?
- we may want to remove the dependency, but it will break the (deprecated) ``kedro lint`` command : it likely needs to wait 0.19
- the pre-commit should still use ``blacken-docs`` with a black dependency which we should keep consistent with the ruff formatter version (black 22 for blacken-docs, but ruff is on black 23). Consistency is likely not a big deal because it won't create conflicts apart from slight local reformatting if someones copy/paste documentation
- we likely want to change the template created for a new project, but I gues this is out of scope for this PR s ?

**Comment by deepyaman** [comment]
> (Seem that I am not allowed to review, but it looks ok for me).
FYI @astrojuanlu @merelcht @yetudada 

> The are still a lot of mentions of [black all across the codebase](https://github.com/search?q=repo%3Akedro-org%2Fkedro%20black&type=code) which raises a couple of questions / remarks :
> 
> * we want to change the documentation, but I guess this is out of scope for this PR?
> * we may want to remove the dependency, but it will break the (deprecated) `kedro lint` command : it likely needs to wait 0.19
> * we likely want to change the template created for a new project, but I gues this is out of scope for this PR s ?
I think most of this is out of scope for this PR, since focus is on formatting the Kedro repo, not on linting user projects (which, as you say, will be deprecated/handled differently). I did bring up in a comment on the add-ons demo that we can drop Black in favor of Ruff, but that would be done separately. 

> * the pre-commit should still use `blacken-docs` with a black dependency which we should keep consistent with the ruff formatter version (black 22 for blacken-docs, but ruff is on black 23). Consistency is likely not a big deal because it won't create conflicts apart from slight local reformatting if someones copy/paste documentation
Good point.

File: .pre-commit-config.yaml [hunk 2]
```diff
@@ -54,18 +57,12 @@ repos:
       rev: v1.12.1
       hooks:
           - id: blacken-docs
-            additional_dependencies: [black~=22.0]
+            additional_dependencies:
+            - black==23.10.1
```
Comment (deepyaman) [inline .pre-commit-config.yaml:61]:
```suggestion
            - black~=23.0
```
3.7 compatibility...

**Review by merelcht — APPROVED** [review]
LGTM! Thanks @deepyaman

File: kedro/framework/cli/micropkg.py [hunk 3]
```diff
@@ -824,13 +824,10 @@ def _move_package_with_conflicting_name(target: Path, conflicting_name: str):
...
 
-_SourcePathType = Union[Path, List[Tuple[Path, str]]]
-
-
-def _generate_sdist_file(  # noqa: too-many-arguments,too-many-locals
+def _generate_sdist_file(  # noqa: PLR0913,too-many-locals
     micropkg_name: str,
     destination: Path,
-    source_paths: tuple[_SourcePathType, ...],
+    source_paths: tuple[Path, Path, list[tuple[Path, str]]],
```
Comment (Galileo-Galilei) [inline kedro/framework/cli/micropkg.py:830]: I guess the goal to change this is to remove all the #type: ignore below?

File: tests/framework/context/test_context.py (hunk 1 above)
Comment (Galileo-Galilei) [reply tests/framework/context/test_context.py:190]: Don't know the answer, sorry!

**Review by Galileo-Galilei — APPROVED** [review]

**Pull request merged**